libraries/integration/Microsoft.Bot.Builder.Integration.Runtime/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -81,55 +88,44 @@ internal static void AddBotRuntimeStorage(this IServiceCollection services)
+                .Get<List<BotComponentDefinition>>() ?? Enumerable.Empty<BotComponentDefinition>();
 
-                // Iterate through configured components and load each one
-                foreach (BotComponentDefinition component in runtimeSettings.Components)
-                {
-                    component.Load(componentEnumenator, services, configuration);
-                }
-            }
+            var componentEnumerator = new AssemblyBotComponentEnumerator(AssemblyLoadContext.Default);
 
-            foreach (BotComponent component in BuiltInBotComponents.GetComponents())
-            {
-                var componentServices = new ServiceCollection();
+            var exceptions = new List<Exception>();
 
-                component.ConfigureServices(componentServices, configuration, null /*for now*/);
-
-                foreach (var serviceDescriptor in componentServices)
+            // Iterate through configured components and load each one
+            foreach (BotComponentDefinition component in componentDefinitions)
+            {
+                try
                 {
-                    services.Add(serviceDescriptor);
+                    component.Load(componentEnumerator, services, configuration);
+                }
+#pragma warning disable CA1031 // Do not catch general exception types. We want to capture all exceptions from components and throw them all at once.
+                catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                {
+                    exceptions.Add(ex);
                 }
             }
-        }
-
-        internal static void AddBotRuntimeAdapters(this IServiceCollection services, RuntimeSettings runtimeSettings)
-        {
-            const string defaultRoute = "messages";
-
-            // CoreBotAdapter registration
-            services.AddSingleton<CoreBotAdapter>();
-            services.AddSingleton<IBotFrameworkHttpAdapter>(sp => sp.GetRequiredService<CoreBotAdapter>());
-
-            // Needed for SkillsHttpClient which depends on BotAdapter
-            services.AddSingleton<BotAdapter>(sp => sp.GetRequiredService<CoreBotAdapter>()); 
-            
-            // Adapter settings so the default adapter is homogeneous with the configured adapters at the controller / registration level
-            services.AddSingleton(new AdapterSettings() { Route = defaultRoute, Enabled = true, Name = typeof(CoreBotAdapter).FullName });
 
-            // Adapter settings for configurable adapters. Runtime controllers pick up this config to get info on adapters and routes.
-            foreach (var adapterSetting in runtimeSettings.Adapters)
+            if (exceptions.Any())
             {
-                services.AddSingleton(adapterSetting);
+                throw new AggregateException("Encountered exceptions while loading bot components.", exceptions);
```

Comment:
Moving to the pattern that asp.net core uses to load IHostingStartup enhancements

https://github.com/dotnet/aspnetcore/blob/80880d0ede83172aad8b48e8cd341b6a339ea8ff/src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs